sympy/core/kind.py

```diff
@@ -100,6 +100,22 @@ def __new__(cls):
     def __repr__(self):
         return "UndefinedKind"
 
+    @property
+    def add(self):
+        from .add import Add
+        return Add
```

Comment:
I have noticed many times when profiling something in sympy that delayed imports like this can have a measurable performance impact. Is there a way to avoid this e.g. by putting the imports at the bottom of the file.

Reply:
Unfortunately, it raises ImportError due to circular import.

Reply:
If this is going to be called every time we do `a+b` then I think we need to find a way to fix it.

Reply:
It took some fiddling but e.g.:

<details><summary>CLICK ME</summary>
<p>

```diff
diff --git a/sympy/core/add.py b/sympy/core/add.py
index 77d9d49271..e43ae27b11 100644
--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -8,7 +8,6 @@
 from .sympify import _sympify
 from .operations import AssocOp
 from .cache import cacheit
-from .numbers import ilcm, igcd
 from .expr import Expr
 from .kind import KindDispatcher
 
@@ -1225,6 +1224,8 @@ def add(*args, evaluate=False, **kwargs):
     func = selected_kind.add
     return func(*args, **kwargs)
 
+import sympy.core.expr as expr
+expr.Add = Add
 
 from .mul import Mul, _keep_coeff, prod
-from sympy.core.numbers import Rational
+from .numbers import ilcm, igcd, Rational
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index 8e82778c7d..2440574b5c 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -8,7 +8,6 @@
 from .sympify import _sympify, sympify, SympifyError
 from .compatibility import iterable, ordered
 from .singleton import S
-from .kind import UndefinedKind
 from ._print_helpers import Printable
 
 from inspect import getmro
@@ -108,8 +107,6 @@ class Basic(Printable, metaclass=ManagedProperties):
     is_MatAdd = False
     is_MatMul = False
 
-    kind = UndefinedKind
-
     def __new__(cls, *args):
         obj = object.__new__(cls)
         obj._assumptions = cls.default_assumptions
@@ -2055,3 +2052,7 @@ def _make_find_query(query):
     elif isinstance(query, Basic):
         return lambda expr: expr.match(query) is not None
     return query
+
+
+from .kind import UndefinedKind
+Basic.kind = UndefinedKind
diff --git a/sympy/core/expr.py b/sympy/core/expr.py
index 1517356402..a002f49b9e 100644
--- a/sympy/core/expr.py
+++ b/sympy/core/expr.py
@@ -4049,12 +4049,3 @@ def search_element(self, elem):
             elif id(arg) == id(elem):
                 return (i,)
         return None
-
-
-from .mul import Mul
-from .add import Add
-from .power import Pow
-from .function import Function, _derivative_dispatch
-from .mod import Mod
-from .exprtools import factor_terms
-from .numbers import Integer, Rational
diff --git a/sympy/core/exprtools.py b/sympy/core/exprtools.py
index 58914ba83b..10ab7fd12f 100644
--- a/sympy/core/exprtools.py
+++ b/sympy/core/exprtools.py
@@ -1571,3 +1571,7 @@ def _pemexpand(expr):
 
         # mid was an Add that didn't factor successfully
         return _keep_coeff(c, g*l*mid*r)
+
+
+import sympy.core.expr as expr
+expr.factor_terms = factor_terms
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 8c7209763a..d75139361e 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -32,14 +32,12 @@
 
 from typing import Any, Dict as tDict, Optional, Set as tSet, Tuple as tTuple, Union
 
-from .add import Add
 from .assumptions import ManagedProperties
 from .basic import Basic, _atomic
 from .cache import cacheit
 from .compatibility import iterable, is_sequence, as_int, ordered, Iterable
 from .decorators import _sympifyit
 from .expr import Expr, AtomicExpr
-from .numbers import Rational, Float
 from .operations import LatticeOp
 from .rules import Transform
 from .singleton import S
@@ -3353,4 +3351,10 @@ def nfloat(expr, n=15, exponent=False, dkeys=False):
         lambda x: isinstance(x, Function)))
 
 
-from sympy.core.symbol import Dummy, Symbol
+import sympy.core.expr as expr
+expr.Function = Function
+expr._derivative_dispatch = _derivative_dispatch
+
+
+from .add import Add
+from .numbers import Rational, Float
diff --git a/sympy/core/kind.py b/sympy/core/kind.py
index 666854445c..8378ca306a 100644
--- a/sympy/core/kind.py
+++ b/sympy/core/kind.py
@@ -102,7 +102,6 @@ def __repr__(self):
 
     @property
     def add(self):
-        from .add import Add
         return Add
 
     @property
@@ -177,7 +176,6 @@ def __repr__(self):
 
     @property
     def add(self):
-        from .add import Add
         return Add
 
     @property
@@ -416,3 +414,5 @@ def __doc__(self):
             docs.append(s)
 
         return '\n\n'.join(docs)
+
+from .add import Add
diff --git a/sympy/core/mod.py b/sympy/core/mod.py
index b4f3de2523..91a1d20ab3 100644
--- a/sympy/core/mod.py
+++ b/sympy/core/mod.py
@@ -237,3 +237,7 @@ def _eval_is_nonpositive(self):
     def _eval_rewrite_as_floor(self, a, b, **kwargs):
         from sympy.functions.elementary.integers import floor
         return a - b*floor(a/b)
+
+
+import sympy.core.expr as expr
+expr.Mod = Mod
diff --git a/sympy/core/mul.py b/sympy/core/mul.py
index 1c989bb4d5..98dae1c068 100644
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -2099,6 +2099,8 @@ def do(e):
         return e
     return bottom_up(e, do)
 
+import sympy.core.expr as expr
+expr.Mul = Mul
 
 from .numbers import Rational
 from .power import Pow
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index 79f6a7d165..c38ec864aa 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -3954,11 +3954,16 @@ def sympify_complex(a):
 
 converter[complex] = sympify_complex
 
-from .power import Pow, integer_nthroot
+
+import sympy.core.expr as expr
+expr.Integer = Integer
+expr.Rational = Rational
+
 from .mul import Mul
 Mul.identity = One()
 from .add import Add
 Add.identity = Zero()
+from .power import Pow, integer_nthroot
 
 def _register_classes():
     numbers.Number.register(Number)
diff --git a/sympy/core/power.py b/sympy/core/power.py
index 943d49a68b..8c09d3644d 100644
--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -5,8 +5,6 @@
 from .singleton import S
 from .expr import Expr
 from .evalf import PrecisionExhausted
-from .function import (_coeff_isneg, expand_complex, expand_multinomial,
-    expand_mul)
 from .logic import fuzzy_bool, fuzzy_not, fuzzy_and
 from .compatibility import as_int, HAS_GMPY, gmpy
 from .parameters import global_parameters
@@ -1750,8 +1748,12 @@ def power(b, e, evaluate=False, **kwargs):
     func = selected_kind.pow
     return func(b, e, **kwargs)
 
+import sympy.core.expr as expr
+expr.Pow = Pow
 
-from .add import Add
 from .numbers import Integer
+from .add import Add
 from .mul import Mul, _keep_coeff
 from .symbol import Symbol, Dummy, symbols
+from .function import (_coeff_isneg, expand_complex, expand_multinomial,
+    expand_mul)
diff --git a/sympy/core/symbol.py b/sympy/core/symbol.py
index 68a5d000f1..e8ef593b01 100644
--- a/sympy/core/symbol.py
+++ b/sympy/core/symbol.py
@@ -5,7 +5,6 @@
 from .singleton import S
 from .expr import Expr, AtomicExpr
 from .cache import cacheit
-from .function import FunctionClass
 from .kind import NumberKind, UndefinedKind
 from sympy.core.logic import fuzzy_bool
 from sympy.logic.boolalg import Boolean
@@ -888,3 +887,10 @@ def disambiguate(*iter):
             ki = mapping[k][i]
             reps[ki] = Symbol(name, **ki.assumptions0)
     return new_iter.xreplace(reps)
+
+
+import sympy.core.function as function
+function.Symbol = Symbol
+function.Dummy = Dummy
+
+FunctionClass = function.FunctionClass
diff --git a/sympy/logic/boolalg.py b/sympy/logic/boolalg.py
index fbcba1f8d4..0a36f46616 100644
--- a/sympy/logic/boolalg.py
+++ b/sympy/logic/boolalg.py
@@ -4,13 +4,11 @@
 
 from collections import defaultdict
 from itertools import chain, combinations, product
-from sympy.core.add import Add
 from sympy.core.basic import Basic
 from sympy.core.cache import cacheit
 from sympy.core.compatibility import ordered, as_int
 from sympy.core.decorators import sympify_method_args, sympify_return
 from sympy.core.function import Application, Derivative
-from sympy.core.numbers import Number
 from sympy.core.operations import LatticeOp
 from sympy.core.singleton import Singleton, S
 from sympy.core.sympify import converter, _sympify, sympify
@@ -3089,3 +3087,6 @@ def simplify_patterns_xor():
                       And(Lt(a, Max(b, c)), Ge(a, Min(b, c)))),
                      )
     return _matchers_xor
+
+from sympy.core.numbers import Number
+from sympy.core.add import Add
diff --git a/sympy/utilities/iterables.py b/sympy/utilities/iterables.py
index d0a5aa4d97..ce1fee7b68 100644
--- a/sympy/utilities/iterables.py
+++ b/sympy/utilities/iterables.py
@@ -6,7 +6,7 @@
 import random
 from operator import gt
 
-from sympy.core import Basic
+from sympy.core.basic import Basic
 
 # this is the logical location of these functions
 from sympy.core.compatibility import (as_int, is_sequence, iterable, ordered)
```

</p>
</details>

Reply:
This is probably something that we should try to fix systematically by defining a clear ordering of the classes.

Reply:
Perhaps a better fix is to move NumberKind into `expr.py`

Reply:
How can I fix it by moving `NumberKind` to `expr.py`?

Reply:
I resolved it by letting `NumberKind.add = Add` in add.py.